### PR TITLE
EIP98 - Remove state root hash from transaction receipt

### PR DESCRIPTION
**Pull request by gumb0**
Implementing Option 1 of EIP - removing state root hash from Receipt RLP

https://github.com/ethereum/cpp-ethereum/issues/3613
https://github.com/ethereum/EIPs/issues/98

- [x] Save zero hash instead of root state hash
- [x] Remove root state hash instead of saving zero hash
- [x] Implement debug RPC methods (trace, storageRangeAt) without relying on state root in receipt
- [x] Fix tests relying on state root in receipts

### CODE REVIEW

File: libethereum/Client.cpp [hunk 1]
```diff
@@ -775,28 +775,6 @@ Block Client::block(h256 const& _blockHash, PopulationStatistics* o_stats) const
 	}
 }
 
-State Client::state(unsigned _txi, h256 const& _blockHash) const
```
Comment (gumb0) [inline libethereum/Client.cpp]: These were not used

File: libethereum/Block.h [hunk 2]
```diff
@@ -202,10 +202,11 @@ class Block
 	/// Get the bloom filter of a particular transaction that happened in the block.
 	LogBloom const& logBloom(unsigned _i) const { return receipt(_i).bloom(); }
 
-	/// Get the State immediately after the given number of pending transactions have been applied.
+	/// Get the State root hash immediately before after all previous transactions before transaction @a _i have been applied.
```
Comment (pirapira) [inline libethereum/Block.h]: "before after" reads strange.

Comment (gumb0) [inline libethereum/Block.h]: Thanks, fixed

File: libethereum/Executive.cpp [hunk 3]
```diff
@@ -141,6 +141,25 @@ string StandardTrace::json(bool _styled) const
 	return _styled ? Json::StyledWriter().write(m_trace) : Json::FastWriter().write(m_trace);
 }
 
+namespace
+{
+
+State& initIntermediateState(State& _s, Block const& _block, unsigned _txIndex, BlockChain const& _bc)
```
Comment (chriseth) [inline libethereum/Executive.cpp]: Can you name this `o_s` to signal that it will be modified?

File: libethereum/Executive.cpp [hunk 4]
```diff
@@ -158,7 +177,7 @@ Executive::Executive(Block& _s, LastHashes const& _lh, unsigned _level):
 }
 
 Executive::Executive(State& _s, Block const& _block, unsigned _txIndex, BlockChain const& _bc, unsigned _level):
```
Comment (chriseth) [inline libethereum/Executive.cpp]: Also here, this should be named `o_s` - but I see this is not part of the PR.

File: libethereum/State.cpp [hunk 5]
```diff
@@ -539,7 +540,27 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
...
+		TransactionReceipt(startGasUsed + e.gasUsed(), e.logs()) :
+		TransactionReceipt(rootHash(), startGasUsed + e.gasUsed(), e.logs());
+	return make_pair(res, receipt);
+}
+
+void State::executeBlockTransactions(Block const& _block, unsigned _txCount, LastHashes const& _lastHashes, SealEngineFace const& _sealEngine)
+{
+	u256 gasUsed = 0;
+	for (unsigned i = 0; i < _txCount; ++i)
+	{
```
Comment (chriseth) [inline libethereum/State.cpp]: The whole for-loop body should be a helper function also used elsewhere - or are the cases there more complicated?

File: libethereum/TransactionReceipt.cpp [hunk 6]
```diff
@@ -44,7 +54,13 @@ TransactionReceipt::TransactionReceipt(h256 _root, u256 _gasUsed, LogEntries con
 
 void TransactionReceipt::streamRLP(RLPStream& _s) const
 {
-	_s.appendList(4) << m_stateRoot << m_gasUsed << m_bloom;
+	if (m_stateRoot)
```
Comment (chriseth) [inline libethereum/TransactionReceipt.cpp]: Can you document this is the header? Perhaps best to put that into the class description.

File: libweb3jsonrpc/Debug.cpp [hunk 7]
```diff
@@ -136,7 +142,15 @@ Json::Value Debug::debug_storageRangeAt(string const& _blockHashOrNumber, int _t
 		Block block = m_eth.block(blockHash(_blockHashOrNumber));
 
 		unsigned const i = ((unsigned)_txIndex < block.pending().size()) ? (unsigned)_txIndex : block.pending().size();
-		State state = block.fromPending(i);
+		State state(block.state());
+		u256 const rootHash = block.stateRootBeforeTx(_txIndex);
+		if (rootHash)
```
Comment (chriseth) [inline libweb3jsonrpc/Debug.cpp]: I have seen this block of code before :-)

Comment (gumb0) [inline libweb3jsonrpc/Debug.cpp]: Extracted this to common function `dev::eth::createIntermediateState` in State.cpp

File: libethereum/State.cpp (hunk 5 above)
Comment (gumb0) [inline libethereum/State.cpp]: Yeah, the other case - `State::execute` needs to get gas, logs, and return value from `Executive`...
But I've moved the common part (initialize-execute-go-finalize) to separate helper method

File: libethereum/TransactionReceipt.cpp (hunk 6 above)
Comment (gumb0) [inline libethereum/TransactionReceipt.cpp]: Added a comment

File: libethereum/State.cpp [hunk 8]
```diff
@@ -617,3 +639,17 @@ std::ostream& dev::eth::operator<<(std::ostream& _out, State const& _s)
...
+State& dev::eth::createIntermediateState(State& o_s, Block const& _block, unsigned _txIndex, BlockChain const& _bc)
+{
+	o_s = _block.state();
+	u256 const rootHash = _block.stateRootBeforeTx(_txIndex);
+	if (rootHash)
+		o_s.setRoot(rootHash);
+	else
+	{
+		o_s.setRoot(_block.stateRootBeforeTx(0));
+		o_s.executeBlockTransactions(_block, _txIndex, _bc.lastHashes(_block.info().parentHash()), *_bc.sealEngine());
```
Comment (chriseth) [inline libethereum/State.cpp]: Do we need a check for dereferencing sealEngine here?

File: libethereum/Executive.cpp [hunk 9]
```diff
@@ -157,8 +157,8 @@ Executive::Executive(Block& _s, LastHashes const& _lh, unsigned _level):
 {
 }
 
-Executive::Executive(State& _s, Block const& _block, unsigned _txIndex, BlockChain const& _bc, unsigned _level):
-	m_s(_s = _block.fromPending(_txIndex)),
+Executive::Executive(State& o_s, Block const& _block, unsigned _txIndex, BlockChain const& _bc, unsigned _level):
```
Comment (chfast) [inline libethereum/Executive.cpp]: I guess the convention is `io_s`, however I'm not big fun of it.

File: libethereum/State.cpp (hunk 5 above)
Comment (chfast) [inline libethereum/State.cpp]: I wanted to move it to a separate function for a long time.

File: libethereum/State.cpp [hunk 10]
```diff
@@ -539,7 +535,33 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
...
 	}
 
-	return make_pair(res, TransactionReceipt(rootHash(), startGasUsed + e.gasUsed(), e.logs()));
+	TransactionReceipt const receipt = _envInfo.number() >= _sealEngine.chainParams().u256Param("metropolisForkBlock") ?
+		TransactionReceipt(startGasUsed + e.gasUsed(), e.logs()) :
+		TransactionReceipt(rootHash(), startGasUsed + e.gasUsed(), e.logs());
+	return make_pair(res, receipt);
+}
+
+void State::executeTransaction(Executive& _e, Transaction const& _t, OnOpFunc const& _onOp)
```
Comment (chfast) [inline libethereum/State.cpp]: It does not depend on State, can we make it a free function somewhere in Executive.h?

File: libethereum/State.cpp (hunk 8 above)
Comment (chfast) [inline libethereum/State.cpp]: I don't think so. I've spot it before, it should not be a pointer type.

File: test/unittests/libethereum/Block.cpp [hunk 11]
```diff
@@ -59,7 +59,7 @@ BOOST_AUTO_TEST_CASE(bStates)
 		OverlayDB const& genesisDB = genesisBlock.state().db();
 		BlockChain const& blockchain = testBlockchain.interface();
 
-		State stateBofore = testBlockchain.topBlock().state();
+		h256 stateBefore = testBlockchain.topBlock().state().rootHash();
```
Comment (chfast) [inline test/unittests/libethereum/Block.cpp]: Rename to `stateRootBefore`?

File: test/unittests/libethereum/Block.cpp [hunk 12]
```diff
@@ -76,15 +76,15 @@ BOOST_AUTO_TEST_CASE(bStates)
 
 		Block block2 = blockchain.genesisBlock(genesisDB);
 		block2.populateFromChain(blockchain, testBlock.blockHeader().hash());
-		State stateAfterInsert = block2.fromPending(0); //get the state of blockchain on previous block
-		BOOST_REQUIRE(ImportTest::compareStates(stateBofore, stateAfterInsert) == 0);
+		h256 stateAfterInsert = block2.stateRootBeforeTx(0); //get the state of blockchain on previous block
+		BOOST_REQUIRE_EQUAL(stateBefore, stateAfterInsert);
```
Comment (chfast) [inline test/unittests/libethereum/Block.cpp]: Just a notice, `BOOST_REQUIRE` is overused in tests. I think `BOOST_CHECK` is usually better.

Comment (gumb0) [inline test/unittests/libethereum/Block.cpp]: I'd say finer-grained tests with a single BOOST_REQUIRE in each one is better than large tests with a series of BOOST_CHECKs

File: libethereum/State.cpp (hunk 10 above)
Comment (gumb0) [inline libethereum/State.cpp]: Maybe better internal in Executive.cpp then?

File: libethereum/Executive.cpp (hunk 9 above)
Comment (gumb0) [inline libethereum/Executive.cpp]: In some sense it's not really used as input, it is assigned immediately another State in `createIntermediateState`. On the other hand it initializes `m_state` member... Ok, I think I'll rename it

File: libethereum/State.cpp (hunk 10 above)
Comment (chfast) [inline libethereum/State.cpp]: If possible, yes.

File: test/unittests/libethereum/Block.cpp (hunk 12 above)
Comment (chriseth) [inline test/unittests/libethereum/Block.cpp]: Always use CHECK if it is possible to continue the test in case it fails, and only REQUIRE if it would not make sense to continue the test.

Comment (chfast) [inline test/unittests/libethereum/Block.cpp]: I think @gumb0 understand the difference. His point is that in case of failure in CHECK boost will output a long list of errors instead of single one per test.

File: libethereum/State.cpp (hunk 8 above)
Comment (gumb0) [inline libethereum/State.cpp]: Yeah, it's always non-null in practice

File: libethereum/State.cpp (hunk 10 above)
Comment (gumb0) [inline libethereum/State.cpp]: I meant here in State.cpp, sorry, moved it now.

File: test/unittests/libethereum/Block.cpp (hunk 12 above)
Comment (gumb0) [inline test/unittests/libethereum/Block.cpp]: Yeah, here it doesn't make sense to proceed with other checks when the first check fails, they are dependent

**Comment by chfast** [comment]
Great job, as usual!

**Comment by pirapira** [comment]
I was asked to review this, and I intend to read the Metropolis changes in all major clients.  I added this to my pending list with priority H and deadline June 15.
